### ISSUE TITLE
Add FRC examples to vision

### ISSUE THREAD

**Issue by Twinters007** [FRC-PDR/ProgrammingDoneRight#18]

**Comment by TuringsEgo** [comment]
what exactly do you want here? I think what is there now is a pretty good summary of FRC in it

**Comment by Twinters007** [comment]
yeah. I wonder if we should have a guide to setting up vision? We should definitely have guides to using coprocessors I think. I think questions about rPi were asked about 6 billion times on the discord.

**Comment by prajy** [comment]
Like information on how to install the necessary tools? Explaining the difference on the OpenCVTegra on Jetson vs regular OpenCV. And downloading openCV on the pi and how to incorporate it onto robot code (I have heard a lot of these questions). Setting up is one thing, integrating is another big step.

**Comment by TuringsEgo** [comment]
Maybe a setup page that would go over doing roborio vision, co-processor vision and DS vision, then pick the big 3 co-processors (jetson, odroid, pi) and go over how to set them up? The problem with that is that jetson setup changes like every 3 months, so it'd have to be maintained to some degree. Nvidia has a special opencv package.

Then another page about how to communicate vision to the roborio? Networktables vs udp packets sort of thing. 

Thoughts?

**Comment by prajy** [comment]
Yeah that sounds good. The major ways to do vision, the 3 big co processors and integrating them to the robot code with communication. Maybe even a part on what to do with the vision values once received.

**Comment by TuringsEgo** [comment]
> Maybe even a part on what to do with the vision values once received

I think that should be reserved to the control section. Maybe just a hyperlink to the particular page. Once the controller gets the degrees offset or distance, the camera becomes just another sensor, albeit a precise one. Lining up to a target when given the degrees offset is essentially the same thing as making the robot face x degrees from a gyro reading.

**Comment by Twinters007** [comment]
Using pynetworktables to communicate, but link to the pynetworktables rtd for actual use, if you know what I mean

**Comment by TuringsEgo** [comment]
Maybe just have a networking subsection then give explicit examples in the 3 languages?

**Comment by prajy** [comment]
Yeah, explicit examples are good. Also, I just released this: https://www.chiefdelphi.com/forums/showthread.php?t=159004. Its basically a framework where you can customize what you want to do in C++. I'm also working on a GUI where you can select some options and bam it will auto generate some code for you (Like GRIP, except its a lot simpler to use and has some extra features.) The GUI is on JavaFX and the code just is based off using the user input checkboxes and uses a buffered writer to write it based of lots of if statements. Not exactly super efficient but not terrible either. Thoughts?

**Comment by TuringsEgo** [comment]
Should we include a grip tutorial (and other GUI based vision stuff, like your project)? I can't decide. It can be argued both ways. I'll go ahead and write up a rough draft of the setup page and we can go from there.

**Comment by prajy** [comment]
GRIP and other GUIs could be put in as a resources page or something like that. Off topic question here: Is the PDR github organization only for like the founders/maintainers or contributors too?

**Comment by Twinters007** [comment]
@TuringsEgo maybe, but grip lacks a lot. I like using grip just because I can see realtime changes to thresholds, including contour filtering, but it doesn't include anything after you have the contours, and the code is **very** verbose, so I like to test stuff, then implement myself. I think it'd be a good idea to explain that is *is* good for figuring out basic thresholding though, and it does that *really* well.

**Comment by prajy** [comment]
Do you guys think we need a part on what to do after finding the filtered out contour. We went over finding the center from it but maybe we could have stuff on finding the angle and distance to the target. Because this year, just finding the center of the boiler contours will not work. You need to know how far you are so that you can shoot x% power. Also, on the contours page, when we talk about looping for each contour, should we just add in an enhanced for loop to show its the same thing.

**Comment by Twinters007** [comment]
Yeah probably. Little bit of physics/math involved

**Comment by prajy** [comment]
So you guys said GRIP lacks a lot. What do you guys think of this:
A flowchart vision program (similar to the lines with GRIP) but with all the restraints taken off. You can have multiple inputs for multiple cameras, put in the exact parameters you want for each equation, and flowchart it however you want.

**Comment by prajy** [comment]
And it generates code in the end

**Comment by Twinters007** [comment]
Well, the main issue with grip is it doesn't do the last 10% of the work. It'll get the contours, and it'll filter them, but it doesn't find the offset, it doesn't get the moment, etc.. I think what we need is to create a guide that takes grip, explains how to use it to get the main jist of the code, transfer that to real code, and add in the final mmph that you can use on the robot

**Comment by prajy** [comment]
Yeah thats why the program I have been working on was going to generate calculations. I scraped the original checkbox design and now it looking like a flowchart type thing  but it but it should when released give complete customization and versatility with calculations. I want the user to be able to put what the requirement is to start calculating, how many bounding boxes do they want, what they want to do with it, etc
